pkg/sentry/socket/plugin/stack/socket.go

```diff
@@ -0,0 +1,874 @@
+	bytes := src.NumBytes()
+	// If src is empty, let plugin stack handle sending 0-byte data based on protocol.
+	if bytes == 0 {
+		rc := cgo.Sendto(s.fd, 0, 0, sysflags, to)
+		ret, err := translateReturn(rc)
+		return int64(ret), err
+	}
+
+	rw := getReadWriter(s.fd)
+	rw.to = to
+	n, err := src.CopyInTo(t, rw)
+	putReadWriter(rw)
+	return n, err
+}
+
+// SendMsg implements socket.Socket.SendMsg.
+func (s *socketOperations) SendMsg(t *kernel.Task, src usermem.IOSequence, to []byte, flags int, haveDeadline bool, deadline ktime.Time, controlMessages socket.ControlMessages) (int, *syserr.Error) {
+	total := src.NumBytes()
+	nonblock := flags & syscall.MSG_DONTWAIT
+	sysflags := flags | syscall.MSG_DONTWAIT
+
+	n, err := s.send(t, src, to, sysflags)
+
+	if nonblock != 0 {
+		return int(n), syserr.FromError(err)
+	}
+
+	var sent int64
+	for err == nil || err == linuxerr.ErrWouldBlock {
+		if n > 0 {
+			src = src.DropFirst64(n)
+			sent += n
+		}
+
+		if sent == total {
+			return int(total), nil
+		}
+
+		if haveDeadline {
+			err = s.waitEventT(t, waiter.EventOut, deadline)
+		} else {
+			err = s.waitEvent(t, waiter.EventOut)
+		}
+
+		if err != nil {
+			if err == linuxerr.ETIMEDOUT {
+				err = linuxerr.ErrWouldBlock
+			}
+			return int(sent), syserr.FromError(err)
+		}
+
+		n, err = s.send(t, src, to, sysflags)
+	}
+
+	return int(sent), syserr.FromError(err)
+}
+
+// Write implements socket.Socket.Write.
+func (s *socketOperations) Write(ctx context.Context, src usermem.IOSequence, opts vfs.WriteOptions) (int64, error) {
+	bytes := src.NumBytes()
+	// If src is empty, let plugin stack handle sending 0-byte data based on protocol.
+	if bytes == 0 {
+		rc := cgo.Write(s.fd, 0, 0)
+		ret, err := translateReturn(rc)
+		return int64(ret), err
+	}
+
+	rw := getReadWriter(s.fd)
+	n, err := src.CopyInTo(ctx, rw)
+	putReadWriter(rw)
+	if n < bytes && err == nil {
+		return n, linuxerr.ErrWouldBlock
+	}
+	return int64(n), err
+}
+
+// waitEvent implements blocking wait on given events
+// until one of the events have been reported.
+func (s *socketOperations) waitEvent(ctx context.Context, event waiter.EventMask) error {
+	var err error
+	t := ctx.(*kernel.Task)
+	e, ch := waiter.NewChannelEntry(event | waiter.EventErr | waiter.EventHUp)
+	s.EventRegister(&e)
+
+	// It's possible events happens between last check and EventRegister.
+	// If this happens and we don't check readiness again, we would miss
+	// the event and get blocked forever.
+	if s.Readiness(event|waiter.EventErr|waiter.EventHUp) == 0 {
+		err = t.Block(ch)
+	}
+
+	s.EventUnregister(&e)
+	return err
+}
+
+// waitEventT implements blocking wait on given events
+// until one of the events have been reported or timeout based on given deadline.
+func (s *socketOperations) waitEventT(ctx context.Context, event waiter.EventMask, deadline ktime.Time) error {
+	var err error
+	t := ctx.(*kernel.Task)
+	e, ch := waiter.NewChannelEntry(event)
```

Comment:
Should this, like the above function, pass `event | waiter.EventErr | waiter.EventHUp` instead of just `event`?

Reply:
Thanks for pointing out. Yes, it should be the same as above. Sorry for missing that. We have updated it in the new version.